tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -1092,6 +1094,93 @@ LogicalResult ConvertTFLConv2DOp::matchAndRewrite(
   return success();
 }
 
+LogicalResult ConvertTFLConv3DOp::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_conv3d_op = cast<TFL::Conv3DOp>(op);
+  RankedTensorType input_type =
+      tfl_conv3d_op.input().getType().dyn_cast<RankedTensorType>();
+  RankedTensorType filter_type =
+      tfl_conv3d_op.filter().getType().dyn_cast<RankedTensorType>();
+  ShapedType output_type =
+      tfl_conv3d_op.getResult().getType().dyn_cast<ShapedType>();
+
+  if (!input_type | !filter_type || !output_type) {
+    return rewriter.notifyMatchFailure(
+        op, "input/filter/output are not all a ranked tensor");
+  }
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool filter_is_qtype =
+      filter_type.getElementType().isa<mlir::quant::QuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::QuantizedType>();
+
+  if ((input_is_qtype != filter_is_qtype) ||
+      (input_is_qtype != output_is_qtype)) {
+    return rewriter.notifyMatchFailure(op,
+                                       "input/filter/output tensor should be "
+                                       "all quantized or all floating-point");
+  }
+
+  int64_t stride_d = tfl_conv3d_op.stride_d();
+  int64_t stride_h = tfl_conv3d_op.stride_h();
+  int64_t stride_w = tfl_conv3d_op.stride_w();
+  ArrayAttr stride = rewriter.getI64ArrayAttr({stride_d, stride_h, stride_w});
```

Comment:
This feel off: we are creating an ArrayAttr here, just to unpack and repack into an ArrayAttr in convertConv3DCommon. Creating Attributes goes through process of uniquing and is stored lifetime of context. Could these just be passed in as ArrayRef?

Reply:
I haven't noticed the advantages of using ArrayRef before. Is the primary purpose of this to avoid copying objects? These attributes are packed in SmallVector now and passed in as ArrayRef, please check if they match your suggestion. As `getPaddingValuesFromPadType` and `CreateOpAndInfer` requires ArrayAttr as parameter, ArrayRef have to be converted to ArrayAttr when calling them. We can change getPaddingValuesFromPadType to use ArrayRef in follow up.